lib/ansible/plugins/inventory/tower.py

```diff
@@ -0,0 +1,187 @@
+    # If the user supplies '@tower_inventory' as path, the plugin will read from environment variables.
+    no_config_file_supplied = False
+
+    def read_tower_inventory(self, tower_host, tower_user, tower_pass, inventory, verify_ssl=True):
+        if not re.match('(?:http|https)://', tower_host):
+            tower_host = 'https://{tower_host}'.format(tower_host=tower_host)
+        inventory_id = inventory.replace('/', '')
+        inventory_url = '/api/v2/inventories/{inv_id}/script/?hostvars=1&towervars=1&all=1'.format(inv_id=inventory_id)
+        inventory_url = urljoin(tower_host, inventory_url)
+
+        request_handler = Request(url_username=tower_user,
+                                  url_password=tower_pass,
+                                  force_basic_auth=True,
+                                  validate_certs=verify_ssl)
+
+        try:
+            response = request_handler.get(inventory_url)
+        except (ConnectionError, urllib_error.URLError, socket.error, httplib.HTTPException) as e:
+            error_msg = 'Connection to remote host failed: {err}'.format(err=e)
+            # If Tower gives a readable error message, display that message to the user.
+            if callable(getattr(e, 'read', None)):
+                error_msg += ' with message: {err_msg}'.format(err_msg=e.read())
+            raise AnsibleParserError(to_native(error_msg))
+
+        # Attempt to parse JSON.
+        try:
+            return json.loads(response.read())
+        except (ValueError, TypeError) as e:
+            # If the JSON parse fails, print the ValueError
+            raise AnsibleParserError(to_native('Failed to parse json from host: {err}'.format(err=e)))
+
+    def verify_file(self, path):
+        if path.endswith('@tower_inventory'):
+            self.no_config_file_supplied = True
+            return True
+        elif super(InventoryModule, self).verify_file(path):
+            return path.endswith('tower_inventory.yml') or path.endswith('tower_inventory.yaml')
+        else:
+            return False
+
+    def parse(self, inventory, loader, path, cache=True):
+        super(InventoryModule, self).parse(inventory, loader, path)
+        if not self.no_config_file_supplied and os.path.isfile(path):
+            self._read_config_data(path)
+        # Read inventory from tower server.
+        # Note the environment variables will be handled automatically by InventoryManager.
+        inventory = self.read_tower_inventory(self.get_option('host'),
+                                              self.get_option('username'),
+                                              self.get_option('password'),
+                                              self.get_option('inventory_id'),
+                                              verify_ssl=self.get_option('verify_ssl'))
+        # To start with, create all the groups.
+        for group_name in inventory:
+            if group_name != '_meta':
```

Comment:
I'm a little iffy about the ungrouped or all group here. I would skip ungrouped.

Reply:
This one is fine it seems. Our API never gives `ungrouped`, but `inventory.add_host(host_name)` in Ansible Inventory API will automatically add an `ungroup` to the hosts.